### PR TITLE
Rust: Add support for the config

### PR DESCRIPTION
**Pull request by dorezyuk**
The PR adds rust support for the configuration handling. 
The PR goes hand in hand with the pr into everest-core https://github.com/EVerest/everest-core/pull/418

### CODE REVIEW

File: everestrs/everestrs/src/lib.rs [hunk 1]
```diff
@@ -38,6 +39,52 @@ mod ffi {
         data: Vec<u8>,
     }
 
+    /// The possible types a config can have. Note: Naturally this would be am
+    /// enum **with** values - however, cxx can't (for now) map Rusts enums to
+    /// std::variant or union.
+    enum ConfigType {
```
Comment (SirVer) [inline everestrs/everestrs/src/lib.rs:48]: use `serde_codegen::schema::interface::Value` and use Serde do deserialize stuff. That makes this code go away and the other code can be reused.

Comment (SirVer) [reply everestrs/everestrs/src/lib.rs:48]: This means pulling this into everstrs which seems fine.

File: everestrs/everestrs/src/lib.rs [hunk 2]
```diff
@@ -85,9 +132,15 @@ mod ffi {
...
         /// Publishes the given `blob` under the `implementation_id` and `name`.
         fn publish_variable(self: &Module, implementation_id: &str, name: &str, blob: JsonBlob);
+
+        /// Returns the module config from cpp.
+        fn get_module_configs(module_id: &str, prefix: &str, conf: &str) -> Vec<RsModuleConfig>;
+
     }
 }
 
+// type RsModuleConfigs = HashMap<String, HashMap<String, ConfigTypes>>;
```
Comment (SirVer) [inline everestrs/everestrs/src/lib.rs]: delete commented code?

File: everestrs/everestrs/src/lib.rs [hunk 3]
```diff
@@ -287,3 +340,86 @@ impl Runtime {
...
+    fn try_from(value: &Config) -> std::result::Result<Self, Self::Error> {
+        match value {
+            Config::Integer(value) => Ok(*value),
+            _ => Err(()),
+        }
+    }
+}
+
+/// Interface for fetching the configurations through the C++ runtime.
+pub fn get_module_configs() -> HashMap<String, HashMap<String, Config>> {
```
Comment (SirVer) [inline everestrs/everestrs/src/lib.rs:396]: return the Json blob and deseriailze it like we do with everything else. It is not worth to have two different ways of handling deserialization in this module. And as mentioned, the code is in the codegen branch already.

File: everestrs/everestrs_sys/everestrs_sys.cpp [hunk 4]
```diff
@@ -80,11 +81,56 @@ JsonBlob Module::call_command(rust::Str implementation_id, rust::Str name, JsonB
     return json2blob(return_value);
 }
 
+void Module::publish_variable(rust::Str implementation_id, rust::Str name, JsonBlob blob) const {
```
Comment (SirVer) [inline everestrs/everestrs_sys/everestrs_sys.cpp:121]: why was this moved here?

File: everestrs/everestrs/src/lib.rs (hunk 1 above)
Comment (dorezyuk) [reply everestrs/everestrs/src/lib.rs:48]: Are you suggesting to serialize the stuff in the c++ land and then pass json to rust and then deserialize it again?

File: everestrs/everestrs_sys/everestrs_sys.cpp [hunk 5]
```diff
@@ -80,11 +81,56 @@ JsonBlob Module::call_command(rust::Str implementation_id, rust::Str name, JsonB
...
+}
+
 std::unique_ptr<Module> create_module(rust::Str module_id, rust::Str prefix, rust::Str conf) {
     return std::make_unique<Module>(std::string(module_id), std::string(prefix), std::string(conf));
 }
 
-void Module::publish_variable(rust::Str implementation_id, rust::Str name, JsonBlob blob) const {
-    handle_->publish_var(std::string(implementation_id), std::string(name),
-                         json::parse(blob.data.begin(), blob.data.end()));
+namespace {
```
Comment (SirVer) [inline everestrs/everestrs_sys/everestrs_sys.cpp]: there is already an anonymous namespace at the top of the file. Just keep one please.

File: everestrs/everestrs_sys/everestrs_sys.hpp [hunk 6]
```diff
@@ -4,11 +4,16 @@
 #include <framework/runtime.hpp>
 #include <memory>
 #include <string>
+#include <utils/types.hpp>
 
 #include "cxxbridge/rust.h"
+// #include "cxxbridge/lib.rs.h"
```
Comment (SirVer) [inline everestrs/everestrs_sys/everestrs_sys.hpp]: remove commented code?

File: everestrs/everestrs/src/lib.rs (hunk 1 above)
Comment (SirVer) [reply everestrs/everestrs/src/lib.rs:48]: actually I misunderstood the C++ stuff - I was assuming that json values are passed around, but for Config that is not the case. retracted.

File: everestrs/everestrs_sys/everestrs_sys.cpp (hunk 4 above)
Comment (dorezyuk) [reply everestrs/everestrs_sys/everestrs_sys.cpp:121]: To keep all impls for Module's functions together

File: everestrs/everestrs/src/lib.rs [hunk 7]
```diff
@@ -287,3 +340,86 @@ impl Runtime {
...
+/// [ffi::ConfigField] and [ffi::ConfigType].
+#[derive(Debug)]
+pub enum Config {
+    Boolean(bool),
+    String(String),
+    Number(f64),
+    Integer(i64),
+}
+
+impl TryFrom<&Config> for bool {
```
Comment (SirVer) [inline everestrs/everestrs/src/lib.rs:355]: Why is this not using everestrs::Error? also below.

File: everestrs/everestrs/src/lib.rs (hunk 3 above)
Comment (SirVer) [reply everestrs/everestrs/src/lib.rs:396]: retracted.

File: everestrs/everestrs/src/lib.rs [hunk 8]
```diff
@@ -287,3 +340,86 @@ impl Runtime {
...
+        .map(|ffi::RsModuleConfig { module_name, data }| {
+            let current = data
+                .into_iter()
+                .map(|field| {
+                    let value = match field.config_type {
+                        ffi::ConfigType::Boolean => Config::Boolean(field.bool_value),
+                        ffi::ConfigType::String => Config::String(field.string_value),
+                        ffi::ConfigType::Number => Config::Number(field.number_value),
+                        ffi::ConfigType::Integer => Config::Integer(field.integer_value),
+                        _ => panic!("This should never happen"),
```
Comment (SirVer) [inline everestrs/everestrs/src/lib.rs]: not a helpful error message. Please include the unexpected value. Or use `unreachable!`.

File: everestrs/everestrs/src/lib.rs [hunk 9]
```diff
@@ -287,3 +340,86 @@ impl Runtime {
...
+pub fn get_module_configs() -> HashMap<String, HashMap<String, Config>> {
+    let args: Args = argh::from_env();
+    let raw_config = ffi::get_module_configs(
+        &args.module,
+        &args.prefix.to_string_lossy(),
+        &args.conf.to_string_lossy(),
+    );
+
+    // Map the nested Vec's into a nested HashMaps.
+    raw_config
```
Comment (SirVer) [inline everestrs/everestrs/src/lib.rs]: this is very hard to read. Can you instead `push` into a named vector so the stuff is easier to read.

File: everestrs/everestrs_sys/everestrs_sys.cpp [hunk 10]
```diff
@@ -80,11 +81,56 @@ JsonBlob Module::call_command(rust::Str implementation_id, rust::Str name, JsonB
...
+
+inline ConfigField get_config_field(const std::string& _name, bool _value) {
```
Comment (SirVer) [inline everestrs/everestrs_sys/everestrs_sys.cpp]: Very clever, but please add a comment what you are doing here.

File: everestrs/everestrs_sys/everestrs_sys.cpp (hunk 5 above)
Comment (dorezyuk) [reply everestrs/everestrs_sys/everestrs_sys.cpp]: I wanted to keep these functions close the the variant - so they are its immediate apparent why they are here. If you feel strange about 2 namespaces - we can remove it and make the functions static. But I feel the closer these functions are to the variant the easier its to maintain and understand

File: everestrs/everestrs/src/lib.rs (hunk 8 above)
Comment (dorezyuk) [reply everestrs/everestrs/src/lib.rs]: It should never happen however if c++ casts an invalid value to enum class it should be UB and I'm not sure what then happens here - will print the value

File: everestrs/everestrs_sys/everestrs_sys.cpp (hunk 5 above)
Comment (SirVer) [reply everestrs/everestrs_sys/everestrs_sys.cpp]: I feel really juck about two anonymous namespaces in one file and static functions are even worse. I appreciate the argument of keeping these close to their use, but I still think it better to move them into the anonymous namespace and only have one area of private stuff in the file. Pretty please with sugar on top? 🐱

**Review by SirVer — APPROVED** [review]
just one more nit, otherwise looks good to me. This in fact looks really nice and seems quite future proof.

**Comment by SirVer** [comment]
@hikinggrass @a-w50 If you still want to review this before we merge it, please speak up. Otherwise we'll probably merge it Friday or monday.

File: everestrs/everestrs_sys/everestrs_sys.cpp [hunk 11]
```diff
@@ -30,6 +33,41 @@ JsonBlob json2blob(const json& j) {
...
+template <typename T, typename... VARIANT_T> struct VariantMemberImpl : public std::false_type {};
+
+template <typename T, typename... VARIANT_T>
+struct VariantMemberImpl<T, std::variant<VARIANT_T...>> : public std::disjunction<std::is_same<T, VARIANT_T>...> {};
+
+/// @brief Static checker if the type T can be converted to `ConfigEntry`.
+///
+/// We use this to detect `get_config_field` overloads which receive arguments
+/// which aren't part of our `ConfigEntry` variant.
+template <typename T> struct ConfigEntryMember : public VariantMemberImpl<T, ConfigEntry> {};
```
Comment (a-w50) [inline everestrs/everestrs_sys/everestrs_sys.cpp:49]: in line with the c++ standard, I would call the struct `is_config_entry_member` or something similar

File: everestrs/everestrs_sys/everestrs_sys.cpp [hunk 12]
```diff
@@ -30,6 +33,41 @@ JsonBlob json2blob(const json& j) {
...
+
+inline ConfigField get_config_field(const std::string& _name, bool _value) {
```
Comment (a-w50) [inline everestrs/everestrs_sys/everestrs_sys.cpp:51]: i would probably have created a single class with visitor functions, which gets the name as its constructor argument and whose instance could then be used directly within `std::visit`

File: everestrs/everestrs_sys/everestrs_sys.cpp (hunk 11 above)
Comment (a-w50) [reply everestrs/everestrs_sys/everestrs_sys.cpp:49]: furthermore, if we're using `std::visit` we got already type checking, so if we don't implement the visitor for a type, it doesn't exist. Shouldn't be that enough?  Of course there might be still implicit conversions, but that might still happen here too

File: everestrs/everestrs_sys/everestrs_sys.cpp [hunk 13]
```diff
@@ -80,11 +118,36 @@ JsonBlob Module::call_command(rust::Str implementation_id, rust::Str name, JsonB
...
+
+    // Iterate over all modules stored in the module_config.
+    for (const auto& mm : module_configs) {
+        RsModuleConfig mm_out{mm.first, {}};
+        mm_out.data.reserve(mm.second.size());
+
+        // Iterate over all configs stored in the mm (our current module).
+        for (const auto& cc : mm.second) {
+            mm_out.data.emplace_back(
+                std::visit([&](auto&& _value) { return ::get_config_field(cc.first, _value); }, cc.second));
```
Comment (a-w50) [inline everestrs/everestrs_sys/everestrs_sys.cpp:147]: Unfortunately it is that painful to convert a c++ variant to a rust enum ...

File: include/utils/config.hpp [hunk 14]
```diff
@@ -127,7 +127,7 @@ class Config {
     ///
     /// \brief checks if the config contains the given \p module_id
     ///
-    bool contains(const std::string& module_id);
+    bool contains(const std::string& module_id) const;
```
Comment (a-w50) [inline include/utils/config.hpp:130]: :+1:

**Review by a-w50 — APPROVED** [review]
@SirVer @dorezyuk :+1: see my personal opinion on the visitor and the type trait.  Feel free to change it or not.

File: everestrs/everestrs_sys/everestrs_sys.cpp (hunk 11 above)
Comment (dorezyuk) [reply everestrs/everestrs_sys/everestrs_sys.cpp:49]: The everest projects are using CamelCase and not snake_case. 
As for why its there - it's primary task it to fail the compilation if some one removes an entry from std::variant - in this case the dev is "politely" hinted that there might be some cleanup to do in the rust-glue code

Comment (a-w50) [reply everestrs/everestrs_sys/everestrs_sys.cpp:49]: I see, so in case we remove one type from the variant, one of the static asserts should fail, although they will never be called then anymore.  A single `std::is_same_v<ConfigEntry, std::variant<currently_known_types...>>` might also work here

File: everestrs/everestrs_sys/everestrs_sys.cpp (hunk 13 above)
Comment (SirVer) [reply everestrs/everestrs_sys/everestrs_sys.cpp:147]: not so surprising, since they have a different memory layout internally. I assume cxx.rs will eventually add code to make this easier, but for now it is what it is.
